# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

In [4]:
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from path import Path
from sklearn.preprocessing import LabelEncoder, StandardScaler

## Read the CSV and Perform Basic Data Cleaning

In [5]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [6]:
# Create our features
X = df.copy()
X.drop("loan_status", axis=1, inplace=True)



X.head()



loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d pymnt_plan    dti  delinq_2yrs  ...  \
0     Source Verified  Mar-2019          n  27.24          0.0  ...   
1            Verified  Mar-2019          n  20.23          0.0  ...   
2            Verified  Mar-2019          n  24.26          0.0  ...   
3            Verified  Mar-2019          n  31.44          0.0  ...   
4        Not Verified  Mar-2019          n  18.76          0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
0          65687.0           38199.0          2000.0   
1         271427.0           60641.0         41200.0   
2          60644.0           45684.0          7500.0   
3          99506.0           68784.0         19700.0   
4         219750.0           25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 85 columns]

In [7]:
#change home_ownership columns to integer values 
label_encoder = LabelEncoder()
label_encoder.fit(X["home_ownership"])
list(label_encoder.classes_)
X["home_ownership"] = label_encoder.transform(X["home_ownership"])
X.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
0    10500.0    0.1719       375.35               3     66000.0   
1    25000.0    0.2000       929.09               1    105000.0   
2    20000.0    0.2000       529.88               1     56000.0   
3    10000.0    0.1640       353.55               3     92000.0   
4    22000.0    0.1474       520.39               1     52000.0   

  verification_status   issue_d pymnt_plan    dti  delinq_2yrs  ...  \
0     Source Verified  Mar-2019          n  27.24          0.0  ...   
1            Verified  Mar-2019          n  20.23          0.0  ...   
2            Verified  Mar-2019          n  24.26          0.0  ...   
3            Verified  Mar-2019          n  31.44          0.0  ...   
4        Not Verified  Mar-2019          n  18.76          0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
0          65687.0           38199.0          2000.0   
1         271427.0           60641.0         41200.0   
2          60644.0           45684.0          7500.0   
3          99506.0           68784.0         19700.0   
4         219750.0           25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 85 columns]

In [8]:
#change verification_status columns to integer values 
#its not a binary value, there are multiple values. so 0/1 encoding would not be best here. 
label_encoder = LabelEncoder()
label_encoder.fit(X["verification_status"])
list(label_encoder.classes_)
X["verification_status"] = label_encoder.transform(X["verification_status"])
X.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
0    10500.0    0.1719       375.35               3     66000.0   
1    25000.0    0.2000       929.09               1    105000.0   
2    20000.0    0.2000       529.88               1     56000.0   
3    10000.0    0.1640       353.55               3     92000.0   
4    22000.0    0.1474       520.39               1     52000.0   

   verification_status   issue_d pymnt_plan    dti  delinq_2yrs  ...  \
0                    1  Mar-2019          n  27.24          0.0  ...   
1                    2  Mar-2019          n  20.23          0.0  ...   
2                    2  Mar-2019          n  24.26          0.0  ...   
3                    2  Mar-2019          n  31.44          0.0  ...   
4                    0  Mar-2019          n  18.76          0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
0          65687.0           38199.0          2000.0   
1         271427.0           60641.0         41200.0   
2          60644.0           45684.0          7500.0   
3          99506.0           68784.0         19700.0   
4         219750.0           25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 85 columns]

In [9]:
#change issue_d columns to integer values 
#its not a binary value, there are multiple values. so 0/1 encoding would not be best here. 
label_encoder = LabelEncoder()
label_encoder.fit(X["issue_d"])
list(label_encoder.classes_)
X["issue_d"] = label_encoder.transform(X["issue_d"])
X.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
0    10500.0    0.1719       375.35               3     66000.0   
1    25000.0    0.2000       929.09               1    105000.0   
2    20000.0    0.2000       529.88               1     56000.0   
3    10000.0    0.1640       353.55               3     92000.0   
4    22000.0    0.1474       520.39               1     52000.0   

   verification_status  issue_d pymnt_plan    dti  delinq_2yrs  ...  \
0                    1        2          n  27.24          0.0  ...   
1                    2        2          n  20.23          0.0  ...   
2                    2        2          n  24.26          0.0  ...   
3                    2        2          n  31.44          0.0  ...   
4                    0        2          n  18.76          0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
0          65687.0           38199.0          2000.0   
1         271427.0           60641.0         41200.0   
2          60644.0           45684.0          7500.0   
3          99506.0           68784.0         19700.0   
4         219750.0           25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 85 columns]

In [10]:
X = pd.get_dummies(X, columns=["pymnt_plan", "hardship_flag", "debt_settlement_flag"])


In [11]:
X.tail()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
68812    10000.0    0.1502       346.76               3     26000.0   
68813    12000.0    0.2727       368.37               3     63000.0   
68814     5000.0    0.1992       185.62               1     52000.0   
68815    40000.0    0.0646      1225.24               1    520000.0   
68816    16000.0    0.1131       350.36               1     72000.0   

       verification_status  issue_d    dti  delinq_2yrs  inq_last_6mths  ...  \
68812                    1        1   9.60          0.0             0.0  ...   
68813                    0        1  29.07          0.0             0.0  ...   
68814                    1        1  14.86          0.0             0.0  ...   
68815                    2        1   9.96          0.0             1.0  ...   
68816                    2        1   7.02          2.0             0.0  ...   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
68812               0.0                   0.0        0.0          20625.0   
68813               0.0                   0.0        0.0          87939.0   
68814               0.0                   1.0        0.0          30592.0   
68815              12.5                   0.0        0.0        1033574.0   
68816               0.0                   1.0        0.0         251486.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
68812            6798.0         11300.0                      5425.0   
68813           60350.0         13500.0                     62939.0   
68814           18611.0          3600.0                     18492.0   
68815           95958.0        100800.0                     78634.0   
68816           74835.0         23000.0                     63090.0   

       pymnt_plan_n  hardship_flag_N  debt_settlement_flag_N  
68812             1                1                       1  
68813             1                1                       1  
68814             1                1                       1  
68815             1                1                       1  
68816             1                1                       1  

[5 rows x 85 columns]

In [12]:
# Create our target
y = df.copy()
y = df.drop(df.columns.difference(['loan_status']), 1)
y.head()

loan_status
0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk

In [13]:

X = pd.get_dummies(X)

In [14]:
X.describe()

loan_amnt      int_rate   installment  home_ownership    annual_inc  \
count  68817.000000  68817.000000  68817.000000    68817.000000  6.881700e+04   
mean   16677.594562      0.127718    480.652863        1.812779  8.821371e+04   
std    10277.348590      0.048130    288.062432        0.941313  1.155800e+05   
min     1000.000000      0.060000     30.890000        0.000000  4.000000e+01   
25%     9000.000000      0.088100    265.730000        1.000000  5.000000e+04   
50%    15000.000000      0.118000    404.560000        1.000000  7.300000e+04   
75%    24000.000000      0.155700    648.100000        3.000000  1.040000e+05   
max    40000.000000      0.308400   1676.230000        3.000000  8.797500e+06   

       verification_status       issue_d           dti   delinq_2yrs  \
count         68817.000000  68817.000000  68817.000000  68817.000000   
mean              0.669994      0.805542     21.778153      0.217766   
std               0.719105      0.714932     20.199244      0.718367   
min               0.000000      0.000000      0.000000      0.000000   
25%               0.000000      0.000000     13.890000      0.000000   
50%               1.000000      1.000000     19.760000      0.000000   
75%               1.000000      1.000000     26.660000      0.000000   
max               2.000000      2.000000    999.000000     18.000000   

       inq_last_6mths  ...  total_il_high_credit_limit  pymnt_plan_n  \
count    68817.000000  ...                6.881700e+04       68817.0   
mean         0.497697  ...                5.572240e+04           1.0   
std          0.758122  ...                5.095845e+04           0.0   
min          0.000000  ...                1.270000e+02           1.0   
25%          0.000000  ...                2.288000e+04           1.0   
50%          0.000000  ...                4.200000e+04           1.0   
75%          1.000000  ...                7.249900e+04           1.0   
max          5.000000  ...                1.426964e+06           1.0   

       hardship_flag_N  debt_settlement_flag_N  initial_list_status_f  \
count          68817.0                 68817.0           68817.000000   
mean               1.0                     1.0               0.123879   
std                0.0                     0.0               0.329446   
min                1.0                     1.0               0.000000   
25%                1.0                     1.0               0.000000   
50%                1.0                     1.0               0.000000   
75%                1.0                     1.0               0.000000   
max                1.0                     1.0               1.000000   

       initial_list_status_w  next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.876121               0.383161               0.616839   
std                 0.329446               0.486161               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 1.000000               0.000000               0.000000   
50%                 1.000000               0.000000               1.000000   
75%                 1.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       application_type_Individual  application_type_Joint App  
count                 68817.000000                68817.000000  
mean                      0.860340                    0.139660  
std                       0.346637                    0.346637  
min                       0.000000                    0.000000  
25%                       1.000000                    0.000000  
50%                       1.000000                    0.000000  
75%                       1.000000                    0.000000  
max                       1.000000                    1.000000  

[8 rows x 88 columns]

In [15]:
# Check the balance of our target values
y['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [16]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [17]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [18]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [19]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [24]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=1000, random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(n_estimators=1000, random_state=1)

In [25]:
# Print the imbalanced classification report
y_pred_rf = brf.predict(X_test)
print(classification_report_imbalanced(y_test, y_pred_rf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.63      0.89      0.06      0.75      0.55       101
   low_risk       1.00      0.89      0.63      0.94      0.75      0.58     17104

avg / total       0.99      0.88      0.64      0.93      0.75      0.58     17205



In [27]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
y_pred = brf.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[   64,    37],
       [ 1954, 15150]], dtype=int64)

In [28]:
# Calculate the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score

balanced_accuracy_score(y_test, y_pred)

0.7597105419148089

In [29]:
y_pred_rf

array(['low_risk', 'low_risk', 'low_risk', ..., 'low_risk', 'low_risk',
       'low_risk'], dtype=object)

In [31]:
# List the features sorted in descending order by feature importance

feat_importances = pd.Series(brf.feature_importances_, index=X.columns)
feat_importances = feat_importances.sort_values(ascending=False)
feat_importances.head()

total_rec_prncp    0.069451
total_rec_int      0.059849
total_pymnt        0.058433
last_pymnt_amnt    0.058058
total_pymnt_inv    0.056632
dtype: float64

### Easy Ensemble Classifier

In [34]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier

eec = EasyEnsembleClassifier(random_state=1)
eec.fit(X_train, y_train) 


EasyEnsembleClassifier(random_state=1)

In [36]:
# Calculated the balanced accuracy score
y_pred = eec.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9215848414822774

In [ ]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred)

In [37]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.06      0.93      0.91      0.11      0.92      0.85       101
   low_risk       1.00      0.91      0.93      0.95      0.92      0.85     17104

avg / total       0.99      0.91      0.93      0.95      0.92      0.85     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    Answer: They both had a similar score at 0.99. 

2. Which model had the best recall score?

     Answer: Easy Ensemble Classifier had a better rec score, at 0.91.  

3. Which model had the best geometric mean score?

     Answer: Easy Ensemble Classifier had a better rec score, at 0.92. 

4. What are the top three features?

     Answer: total_rec_prncp, total_rec_int and total_pymnt had the top 3 features. 